load_balancer/src/load_balancer.py

```diff
@@ -0,0 +1,226 @@
+
+            for index in range(forwardServers.size()):
+                self.serverSockets[index].shutdown(socket.SHUT_RDWR)
+                self.serverSockets[index].close()
+
+            print("~SERVER SHUTDOWN")
+            time_now = time.strftime("%a, %d %b %Y %H:%M:%S", time.localtime())
+            print("Timestamp: ", time_now)
+
+            self.packetQ.join()
+
+        except socket.error as err:
+            pass
+
+    def producerThread(self):
+        """ Listens to the socket for incoming packets. When a incoming connection
+            occurs a thread will be started to recieve and decode the packet.  """
+
+        print("Starting load balancer PRODUCER thread")
+        self.socketRevc.listen(5)
+        
+        while self.serverRunning:
+            (conn, address) = self.socketRevc.accept()
+
+            conn.setblocking(False)
+
+            # Start a thread for each socket connection.
+            # TODO: Change to be Event driven 
+            Thread(target=self.handleConnection, args=(conn, address)).start()
+
+        print("WARNING! Load Balancer producer thread ending")
+
+    def handleConnection(self, conn, address):
+        """ Listens to the socket for incoming packets. When a incoming connection
+            occurs a thread will be started to recieve and decode the packet. 
+            Will add the incoming packets to internal queue for later transmission.
+            """
+
+        print("New Client Connection Detected from: {}".format(address))
+
+        while self.serverRunning:
+            try:
+                data = conn.recv(PACKET_SIZE)
+                if not data:
+                    # connection has been closed
+                    break
+            except socket.error as err:
+                time.sleep(0.001)
+                continue
+
+            strData = str(data).strip("b\"")
+
+            # Sometimes Data packets are bunched up in the read buffer. 
+            #   This mechanism will seperate them, and process each
+            packets = strData.split("}")
+
+            for userReqData in packets[:-1]:
+                userReqData = userReqData.strip() + '}'
+
+                # Decode the packet into a dictionary type
+                userDict = ast.literal_eval(str(userReqData))
+
+                # DEBUG PRINT
+                #print("Producer -- Putting P on Queue")
+
+                self.packetQ.put(userDict)
+
+        print("Client Closed: {}".format(address))
+        conn.close()
+
+
+    def consumerThread(self):
+        """ Consumer thread
+    
+        """
+        print("Starting load balancer CONSUMER thread")
+
+        while self.serverRunning:
+            try: 
+                packet = self.packetQ.get(False)
+
+                if "user" in packet.keys():
+                    user = packet["user"]
+                else:
+                    user = "default"
+                
+                serverHash = hash(user) % NUM_FORWARD_SERVERS
```

Comment:
this is the constant hashing of the balancer right?

Reply:
Yep! 
It's constant for the application runtime. If you restart the LB the same user might be directed to a different server.

Reply:
Ahh nice work! We can test it more exhaustively when we have our apps more clearly separated into services